### ISSUE TITLE
Hook up e2e tests with postgres docker container

### ISSUE THREAD

**Issue by kurtisassad** [hicommonwealth/commonwealth#4577]
## Description
One issue we have when writing e2e tests is environment drift. This is because we always develop with a populated db, but the e2e tests run on an empty db. To lessen this drift we should write some code to spin up and connect to a pgsql docker container if docker is installed on the machine. This will allow us to test the e2e tests with a fresh db without destroying the developers current db (Like the unit tests).

## Engineering Requirements
- Add support for connecting the e2e tests to a docker container of a postgres image.

**Comment by kurtisassad** [comment]
@jnaviask So I was wondering if we could create a hard dependency between docker and our tests. Basically the most common way to write tests that require outside integration is to use this library:

https://testcontainers.com/

In this library we need to write logic inside the test to startup containers which the tests will connect to. So if we were to go ahead and use this library we would need to either:

a. Create a hard dependency between the tests and docker
b. Write lots of messy custom logic in order to have a localhost fallback if docker is not installed

I would like to use a, since it seems the most clean and commonly used solution. Although we could use b if you think it is worthwhile to still support tests for local infra.

**Comment by jnaviask** [comment]
Testcontainers seems fine as a dependency -- and it's fine if the tests are only runnable via a docker command, as long as there's a way to still debug them (with the browser/console). @kurtisassad

**Comment by kurtisassad** [comment]
Correct, the tests will only be runnable if docker is installed. All debugging will be the same because the app itself is still being run locally (so all breakpoints work + browser).

**Comment by kurtisassad** [comment]
Looking a little more difficult than anticipated. Basically we run into a chicken or the egg scenario:

Normal startup sequence:
`postgresDB -> server -> playwrightTests.`

This PR startup sequence:
`server -> playwrightTests -> postgresDB`

I see two solutions:
1. have wait logic in the server until postgres database is ready
2. Create a dockerfile to build the server, then we just need to build it in its own test container after the pgDB

I am going with 1 since this is the easier option. It will lead to faster but less reliable local tests (possible environment drift).

2. Can be expanded upon in another PR if determined worthwhile.

**Comment by kurtisassad** [comment]
After talking with Ilija, it seems that a better solution is to create a volume containing the test data. So I the startup will look like this:

locally:
1. Run volume builder, this will spin up a docker pgsql container, run db migrations, populate db, then spin down the container. We will be left with a volume that can be used in tests (containing test data).
a. This will comprise of two parts.
2. Run test with copy of the volume. This will start up a pgsql container, attach a copy of the volume, then run the tests.

The issue still remains, if we start up the pgsql testcontainer in the test itself, how do we propagate this information to the server (port info)? It is bad practice to make it predefined. It will also could conflict on ports with the developers machine.

CI:
1. Hash sql population script. Compare with hash in cache
a. If not there, or mismatched hash, rebuild volume and store in cache with hash
b. If there, pull volume from cache
2. Run test with copy of the volume.

**Comment by kurtisassad** [comment]
So I have created diagrams of the two flows I had described during the parking lot:


![Image](https://github.com/hicommonwealth/commonwealth/assets/14794654/401e1ac7-5837-440f-9d04-15fb0a3518c2)


![Image](https://github.com/hicommonwealth/commonwealth/assets/14794654/5bc46d2e-6bf0-4741-8590-bbbdc64655f5)


The problem is wether we want the server to be run locally or in docker. The problem with running it locally (hard solution) is that test-containers which manages the ports and makes sure that the containers are killed after tests, runs per process. Meaning when the pid no longer exists, it will kill the container that started it . I could write custom code to interact with this container (https://hub.docker.com/r/testcontainers/ryuk), but it might not be easy.

This makes it difficult to integrate with the "difficult solution"

The proposed solution has downsides:
- It will take more effort to build the docker container
- It will make local tests more slow as we need to build the docker container first.

Proposed solution is same as standup (run the tests inside the commonwealth container).

**Comment by jnaviask** [comment]
@kurtisassad why can't we just have the commonwealth container depend on (/ inherit) the postgres container, so we don't need to pass around URLs?

**Comment by kurtisassad** [comment]
Although this way solves this whole issue, It is a bad pattern as talked about in the testcontainers docs:
https://node.testcontainers.org/features/containers/#exposing-container-ports

> Testcontainers will automatically bind an available, random port on the host to each exposed container port. This is to avoid port conflicts when running tests quickly or in parallel.
...
specify fixed host port bindings (not recommended):

Essentially it runs into these problems:
1. How do we ensure the port we choose doesn't conflict with existing applications locally?
2. If in the future we want parallel runs, how can we achieve this?

If you think the tradeoff is worth it, we can do it that way (The draft PR will basically need minimal modification)

**Comment by kurtisassad** [comment]
Ok after parking lot, we decided to containerize application. We will take the proposed solution.

Basically in tests, the globalSetup will have:
```
function setup() {
  dbUrl = testContainers.createDbContainer()
  testContainers.createCommonContainer().withEnv(dbUrl)
}
```

Then in the remaining test section, it will run our e2e tests against the commonwealth container.

**Comment by kurtisassad** [comment]
I decided to remove the containerizing the app approach. After getting it working, it basically made tests take ~200-300 more seconds to start up locally, which I find an unacceptable tradeoff. Instead I had started the pgsql container in a separate script, which then calls the test-e2e script and passes the environment variables to them.

**Comment by jnaviask** [comment]
cc @kurtisassad I marked this ticket as "teed up" because the PR is in draft -- can you label "blocked" if blocked + explain why?

**Comment by kurtisassad** [comment]
@jnaviask Finished and ready to go

**Comment by kurtisassad** [comment]
Ok so adding blocked. Ilija talked about wanting to revamp the way we handle these e2e tests, so this should probably be closed

**Comment by kurtisassad** [comment]
Decided to revamp the e2e tests, so this change is no longer necessary
